samples/matter/common/src/matter_init.cpp

```diff
@@ -0,0 +1,238 @@
+	}
+
+#if defined(CONFIG_NET_L2_OPENTHREAD)
+	sInitResult = ThreadStackMgr().InitThreadStack();
+	VerifyInitResultOrReturn(sInitResult, "ThreadStackMgr().InitThreadStack() failed");
+
+	sInitResult = ConfigureThreadRole();
+	VerifyInitResultOrReturn(sInitResult, "Cannot configure Thread role");
+
+#elif defined(CONFIG_CHIP_WIFI)
+	if (!sCommissioningInstance) {
+		sInitResult = CHIP_ERROR_INTERNAL;
+		VerifyInitResultOrReturn(sInitResult, "No valid commissioning instance");
+	}
+	sCommissioningInstance->Init();
+#else
+	sInitResult = CHIP_ERROR_INTERNAL;
+	VerifyInitResultOrReturn(sInitResult, "No valid L2 network backend selected");
+#endif /* CONFIG_NET_L2_OPENTHREAD */
+
+#ifdef CONFIG_CHIP_OTA_REQUESTOR
+	/* OTA image confirmation must be done before the factory data init. */
+	OtaConfirmNewImage();
+#endif
+
+#ifdef CONFIG_MCUMGR_TRANSPORT_BT
+	/* Initialize DFU over SMP */
+	GetDFUOverSMP().Init();
+	GetDFUOverSMP().ConfirmNewImage();
+#endif
+
+	/* Initialize CHIP server */
+#if CONFIG_CHIP_FACTORY_DATA
+	if (sFactoryDataProvider) {
+		sInitResult = sFactoryDataProvider->Init();
+		VerifyInitResultOrReturn(sInitResult, "FactoryDataProvider::Init() failed");
+	}
+
+	SetDeviceInstanceInfoProvider(sFactoryDataProvider);
+	SetDeviceAttestationCredentialsProvider(sFactoryDataProvider);
+	SetCommissionableDataProvider(sFactoryDataProvider);
+#else
+	SetDeviceInstanceInfoProvider(&DeviceInstanceInfoProviderMgrImpl());
+	SetDeviceAttestationCredentialsProvider(Examples::GetExampleDACProvider());
+	/* The default CommissionableDataProvider is set internally in the GenericConfigurationManagerImpl::Init(). */
+#endif
+
+	VerifyOrReturn(sServerInitParams, LOG_ERR("No valid server initialization parameters"));
+	sInitResult = sServerInitParams->InitializeStaticResourcesBeforeServerInit();
+	VerifyInitResultOrReturn(sInitResult, "InitializeStaticResourcesBeforeServerInit() failed");
+
+	sInitResult = PlatformMgr().AddEventHandler(reinterpret_cast<PlatformManager::EventHandlerFunct>(arg), 0);
+	VerifyInitResultOrReturn(sInitResult, "Cannot register CHIP event handler");
+
+	SetDeviceInfoProvider(sDeviceInfoProvider);
+
+	sInitResult = Server::GetInstance().Init(*sServerInitParams);
+	VerifyInitResultOrReturn(sInitResult, "Server::Init() failed");
+
+	ConfigurationMgr().LogDeviceConfig();
+	PrintOnboardingCodes(RendezvousInformationFlags(chip::RendezvousInformationFlag::kBLE));
+	AppFabricTableDelegate::Init();
+
+	if (sPostServerInitClbk) {
+		sInitResult = sPostServerInitClbk();
+		VerifyInitResultOrReturn(sInitResult, "Custom post server initialization failed");
+	}
+}
+
+CHIP_ERROR WaitForReadiness()
+{
+	InitGuard::Wait();
+	return sInitResult;
+}
+} // namespace
+
+/* Public API */
+namespace Nordic
+{
+namespace Matter
+{
+	CHIP_ERROR InitServer(PlatformManager::EventHandlerFunct matterEventHandler, InitData initData)
+	{
+		sCommissioningInstance = initData.mNetworkingInstance;
```

Comment:
You could save the entire `initData` instead of all members separately.

Reply:
Fair point. But if we keep the whole `InitData` structure in the matter_init.cpp, will the default static objects (implementing Matter interfaces) of this instance be stripped by the compiler/linker? It is the case when we initialize the structure immediately (e.g. when passing it by copy to the function) but I wonder if the behavior is the same if we first create the default-initialized object and then overwrite it in `InitServer` function.

Reply:
Good point. I guess you could include `matterEventHandler` as `initData` member and then allocate the copy of `initData` on heap and pass it directly to `ScheduleWork`. But feel free to ignore this comment as it's not that important after all.

Reply:
Or we can initialize the local InitData with `nullptr`s instead of using default initialization, right?

Reply:
Yes, we can.